runtimes/huggingface/setup.py

```diff
@@ -36,6 +36,7 @@ def _load_description() -> str:
     install_requires=[
         "mlserver",
         "optimum[onnxruntime]>=1.4.0, <1.8.0",
+        "tensorflow==2.12.0",
```

Comment:
No, it is not part of the `optimum[onnxruntime]` here is the [link](https://github.com/huggingface/optimum/blob/f7f1ef1d9d8e6fbadb5df9051e4c3685ec4ebc93/setup.py#L43). It is part of some other installations of  [link](https://github.com/huggingface/optimum/blob/f7f1ef1d9d8e6fbadb5df9051e4c3685ec4ebc93/setup.py#L59) of Optimum but not the `optimum[onnxruntime]` that is in MLServer. I spent some time yesterday figuring out the problem and tried testing it locally. The problem is that for some yet unknown reason, having Tensorflow here or in `requirements/dev.txt` will result in `pip` and `tox` to get stuck in [backtracking](https://pip.pypa.io/en/stable/topics/dependency-resolution/#backtracking) for versioning resolution. I've tried a number of changes in the version of the dependencies and still no luck. I think we can do either of the following:
1. Remove support of the Tensorflow for this feature and add a warning message for using requesting it for users trying this feature with Tensorflow (as most of the HF models are PyTroch anyway these days)
2. Continue resolving the TF scenario if you think they both should be supported and partial support for one framework is not acceptable.

Reply:
@adriangonz I added the 1st option, please let me know if there is any problem. I'll look into TensorFlow support in the future.

Reply:
Hey @saeid93 ,

Just as an update, we merged a PR today which should make your life easier on this one:

https://github.com/SeldonIO/MLServer/pull/1092

Essentially, what was going was:

- At the top-level MLServer package, we had to pin `protobuf` to `>=3.20.2` to avoid a CVE with previous versions.
- However, TF `<2.12` is very stubborn on not using that version of `protobuf`. In fact, pip keeps backtracking until it gets to TF `2.9.0` (which mysteriously didn't have that `protobuf` constrain, so is happy).
- This leads to MLServer indirectly forcing an older version of TF which was incompatible with Alibi-Detect and a few others. 
- The fix has been to move that `protobuf` constrain to the official Docker images - letting the user control what goes into their local environment (which could include the faulty version of `protobuf`).

Based on the above, what I would try is to rebase from `master` and then remove the extra `tensorflow` dep in the `runtimes/huggingface/setup.py`. That should hopefully let `huggingface > transformers` pick up the correct version of TF.

Reply:
Hey @adriangonz ,
I figured out the structure of the test and tox in the project by looking at the `setup.cfg` file. I think the best place to add the `tensorflow` is the `setup.py` of huggingface as adding it the `requirements/dev.txt` will install `tensorflow` unnecessarily in other tests. Also, since we now have the  `import tensorflow as tf` in the `runtimes/huggingface/mlserver_huggingface/common.py` I think it makes sense to install it as a dependency. I checked several versions and the one that I have put a cap is the one without conflicts (there were still some conflicts other than the `protobuf`) and it won't get stuck in the backtracking anymore.
Please let me know if you see any remaining issues.

Reply:
Hey @adriangonz,

Thank you for the description. I tested several times and highly doubt that it will be installed automatically or there is any problem with my environment. Are yo sure that `tensorflow` is not installed from before on your environment? Because I checked with several fresh envs and also MLServer Huggingface tox, tensorflow was not installed automatically in both of them.

In the transformers [`setup.py`](https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/setup.py#L103) the `_deps` list is only used for making as a reference for finding the correct version of each package see [here](https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/setup.py#L197). The resulting `deps` dictionary after performing the regular expression maps each package name to its version constraint. Not everything in that list will be necessarily installed. What is installed in the transformers is the [`install_requires`](https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/setup.py#L413) list that is composed of `deps` that are made [above](https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/setup.py#L216).

Another sign for this is that if you see the same [optimum code that you sent me](https://github.com/huggingface/optimum/blob/4fafbb1d4a7cfbb98ddf026e0f05e93ef0d4a2cb/setup.py#L18) they have also pinned [PyTorch](https://github.com/huggingface/optimum/blob/4fafbb1d4a7cfbb98ddf026e0f05e93ef0d4a2cb/setup.py#L19) there although [it is present](https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/setup.py#L180) in the same `_deps` list that you mentioned in the `setup.py` repository. If the `PyTorch` was automatically installed then there was no need to add it there in the `Optimum` library.

If you think pinning `tensorflow` could cause problems, maybe we can revert back to not supporting `tensorflow` or any other options you think might be useful. I'm happy to investigate further if you have any suggestions.

Reply:
@adriangonz Sure, glad to be of any help. I loosened the tf version, however, it will search until gets to the version that I had previously fixed. Please have a look now.

Reply:
Yuup, it seems that `protobuf` is still interacting with the version of TF even though it's not pinned by the main `mlserver` package anymore. I think it's OK if it still defaults to TF 2.9 though - not much we can do there (pinning it could cause plenty of incompatibilities with the other runtime deps). 

So as it is right now (i.e. without any pin), it's probably as good as it gets!